### DISCUSSION TITLE
Inquiry on how to go about improving Japanese parsing quality

### DISCUSSION THREAD

**Discussion by nlovell1** [explosion/spaCy#6749] [Help: Coding & Implementations]
Hi! I'm a newbie to NLP and the whole world of language as it relates to technology, so there are probably a lot of gaps in my knowledge. But, here is my question.

Essentially, I am an avid user of [Morphman](https://github.com/kaegi/MorphMan) , which has recently been experimenting with using Spacy's models to parse sentences so that language sentence flashcards can be reordered such that there is only one new vocabulary or grammar structure used.

In the example of 気になった、 for example, it seems that it is parsed [('気', 'NOUN'), ('に', 'ADP'), ('なっ', 'VERB'), ('た', 'AUX'), ('。', 'PUNCT')]. While the model does extremely well catching exactly how the morphemes relate to one another, this type of breakdown is not helpful for the learner. With consideration to JMDict (primary dictionary for english-Japanese), I imagine it would be parsed as something like 気になっ (VERB) た (AUX). I know I can make exceptions in the parsing, but I am wondering if there is a way to brute force the entries of JMDict to combine with the great parsing going on here to output more semantically descriptive results.

Additionally, has there been any interest to train a model that works well for spoken language yet?

Thanks a ton

**Comment by polm** [comment]
Given what you want to do, I would recommend running JMDict entries through spaCy and using them to generate Matcher rules. Then you could have a "JMDict entry" entity and extract that. 

> I imagine it would be parsed as something like 気になっ (VERB) た (AUX).

You seem to be suggesting you want to change the output of the tokenizer and POS tagger, but that doesn't really make sense. What the Entity Matcher will do is give you Spans, which are lists of tokens in the sentence, which you can turn into single strings if you need to.

**Reply by nlovell1** [reply]
Thanks for your comment. I checked that out yesterday, seems to be a very powerful tool. However, I imagine this sort of issue coming into play-
The dictionary entry in JM dict for what I mentioned above is 好きになる, for example. If what I want to parse ends up being 好きになった, it wouldn't be able to match- I imagine a work around would be to handcode a conjugation reference table (not sure if that's the name) for JMdict entries, but I imagine that would be quite cumbersome, and add a lot of strain especially with speed as a concern (the Dictionary is over 100K entries)!

I'm considering stripping down the dictionary to only 'expressions', or things like 好きになる that have many parts
It's worth noting that some expressions only appear in the negative conjugation when a verb/adj attaches at the end, and it would be a lot to hand search through the entries and hard code them all. Do you know of a workaround for this?
Thanks again

**Reply by polm** [reply]
Just use the `LEMMA` feature and it will match fine.

```
from spacy.matcher import Matcher

ja = spacy.load('ja_core_news_sm')

pattern = [{'LEMMA': '気'}, {'LEMMA': 'に'}, {'LEMMA': 'なる'}]

matcher = Matcher(ja.vocab)
matcher.add("kininaru", None, pattern)

for text in ('あの子が気になる', 'あれが気になった'):
    doc = ja(text)
    matches = matcher(doc)
    _, start, finish = matches[0]
    print(doc[start:finish])
```

Output is:

    気になる
    気になっ

**Reply by nlovell1** [reply]
Wow! Looks great. Thank you very much. If I wanted the expected output to be
あの　子　が　気になる

あれ　が　気になっ　た

So what you suggested (printing the list of tokens in the sentence) should account for the matching how I want it without trying to change the output of the tokenizer?

**Reply by nlovell1** [reply]
https://spacy.io/usage/processing-pipelines#component-example2

In this example, what exactly is happening? Am I correct in that the morphemizer takes into account the rules when outputting? Is this faster than what you've already suggested?

**Reply by polm** [reply]
Your link uses the PhraseMatcher, which is typically faster but less versatile than the Matcher. I didn't realize this but it looks like you can set the PhraseMatcher to use the lemma attribute, in which case it should work just fine.

https://stackoverflow.com/questions/58134994/spacy-how-to-get-lemma-based-phrasematcher

Details on Matcher vs PhraseMatcher:

https://spacy.io/usage/rule-based-matching#matcher-vs-phrase-matcher

In your comment with example tokenization, you can get that tokenization by merging entities. 

https://spacy.io/api/pipeline-functions#merge_entities

> So what you suggested (printing the list of tokens in the sentence) should account for the matching how I want it without trying to change the output of the tokenizer?

If you merge the entities you can get output like your example without changing the tokenizer. 

Also, I have no idea what you mean by "morphemizer". You also referred to the "parser" in your first post without actually mentioning any dependency parser output. This makes it hard to understand what you mean. See here for names of typical pipeline components:

https://spacy.io/usage/processing-pipelines

**Reply by nlovell1** [reply]
Sorry about the ambiguity in my usage of terms- you've been able to answer my questions quite sufficiently despite the confusion you've mentioned. What is the difference between the solution you've proposed and retraining a model such that its tokenization already has these expressions built-in (i.e. changing the tokenizer) without a need for post processing (entity merging, and Matcher/PhraseMatcher)?

If we do go the post-processing route, what is your suggested method of implementing JMDict's thousands of entries? Is this a viable solution when compared to an already packaged model that (if trained how I want it, pending the answer to my last question) that is able to tokenize to my expected output right away?

Curious on your thoughts.
Thank you!

**Reply by polm** [reply]
>  What is the difference between the solution you've proposed and retraining a model such that its tokenization already has these expressions built-in (i.e. changing the tokenizer) without a need for post processing (entity merging, and Matcher/PhraseMatcher)?

How would you change the tokenizer to work the way you want it to? 

> If we do go the post-processing route, what is your suggested method of implementing JMDict's thousands of entries?

Use [jamdict](https://github.com/neocl/jamdict), iterate over all the entries, filter to select the ones you want (maybe only ones that contain a verb and a noun or something), add rules to a [PhraseMatcher](https://spacy.io/usage/rule-based-matching#phrasematcher) using the LEMMA attribute.

Using the PhraseMatcher is pretty straightforward; changing the tokenizer to get what you want seems like much more work and more prone to cause unexpected errors.

**Reply by nlovell1** [reply]
> How would you change the tokenizer to work the way you want it to?

Quite frankly, I'm not sure. I thought that there was maybe a way to retrain the model/change a parameter of some sort. Maybe with something like Prodigy. But this is probably me just being unfamiliar with how the models work.

Say I am unable to account for all expressions in Japanese. How would my solution be able to extrapolate an expression? Especially since jamdict (as far as I can tell?) is J - E... I imagine a monolingual dictionary would have more entries.

**Reply by polm** [reply]
> Quite frankly, I'm not sure. I thought that there was maybe a way to retrain the model/change a parameter of some sort. Maybe with something like Prodigy. But this is probably me just being unfamiliar with how the models work.

Japanese tokenization is provided by [SudachiPy](https://github.com/WorksApplications/SudachiPy) and isn't trainable in spaCy. Prodigy doesn't have a tokenizer mode; you could make tokenization data by splitting a document into characters and doing per-character annotations but it would be awkward. I am not aware of better tools for making tokenizer training data.

You could add dictionary entries to SudachiPy more easily, but changing the tokenizer directly creates the potential for more errors - what if one expression contains another? You could work around that by calculating term weights, but you'd need training data (sentences) for it. You'd also have to provide your own lemmas, readings, etc.

> Say I am unable to account for all expressions in Japanese. How would my solution be able to extrapolate an expression?

How do you define an "expression"? To me it sounds like the most useful definition for your application is "anything in JMDict". What would you do with an expression you identified that you didn't have a dictionary entry for?

> Especially since jamdict (as far as I can tell?) is J - E... I imagine a monolingual dictionary would have more entries.

JMDict is JE only, there's no similar free monolingual Japanese dictionary.

**Reply by nlovell1** [reply]
> what if one expression contains another?

Could you give an example of this (sorry, can't think of any atm)? And could you explain how/why using PhraseMatcher rules would be, comparatively less error prone than changing the SudachiPy tokenizer?

> How do you define an "expression"? To me it sounds like the most useful definition for your application is "anything in JMDict".

This is a fair definition, at least starting with the entries in the dictionary labeled as 'Expression', perhaps expanding into proper names.

>  What would you do with an expression you identified that you didn't have a dictionary entry for?

This was said under the assumption JMDict didn't have as many expressions in Japanese that exist (reasonably speaking), and that its coverage was limited due to being J-E.

Lastly:
> Use jamdict, iterate over all the entries, filter to select the ones you want (maybe only ones that contain a verb and a noun or something), add rules to a PhraseMatcher using the LEMMA attribute.

This solution seems like the easiest, but wouldn't it be extremely slow? I'd have to create a PhraseMatcher instance for every entry for each of my desired expressions, and then iterate through them to see if the expression is in the inputted sentence(s). Especially given the scale of the dictionary, and the fact that SudachiPy is quite slow (compared to something like MeCab, for example), wouldn't you think that this might take a while?

Thanks for your continued interest.

**Reply by polm** [reply]
> This solution seems like the easiest, but wouldn't it be extremely slow? I'd have to create a PhraseMatcher instance for every entry for each of my desired expressions, and then iterate through them to see if the expression is in the inputted sentence(s). 

You can (should) make one PhraseMatcher instance with multiple patterns. 

> Especially given the scale of the dictionary, and the fact that SudachiPy is quite slow (compared to something like MeCab, for example), wouldn't you think that this might take a while?

SudachiPy is slower than MeCab but it's not *that* slow. 10k short dictionary entries won't take 6 hours to process or anything, and you can serialize the PhraseMatcher after creating it.

**Reply by nlovell1** [reply]
```
import spacy
from spacy.matcher import PhraseMatcher

ja = spacy.load("ja_core_news_lg")
matcher = PhraseMatcher(ja.vocab, attr="LEMMA")
matcher.add("EXPRESSION", None, ja("気になる"))
matcher.add("EXPRESSION", None, ja("手も足も出ない"))
matcher.add("EXPRESSION", None, ja("んだ"))

for text in ('あの子が気になる', 'あれが気になった', 'あなたは手も足も出ないんだぞ'):
     doc = ja(text)
     matches = matcher(doc)
     _, start, finish = matches[0]
     print(doc[start:finish])

```
Here is some code that I came up with. 
1.  I noticed that once there is a match found in a string, it goes to the next one.
2. I tried finding examples of merging matched entities and printing them out, but I wasn't able to find any. The Spacy doc page you linked wasn't clear to me. Do you have any other alternative resources I could take a look at?
3. [This user](https://stackoverflow.com/questions/63094439/matching-multi-token-entities-in-spacy) used the EntityRuler to accomplish a similar goal. What would be the difference between that and what I'm doing with PhraseMatcher?
4. How would PhraseMatcher handle an expression within an expression? Would it just default to the larger one?

**Reply by polm** [reply]
1. What's the question?
2. The docs for [the retokenizer](https://spacy.io/api/doc#retokenize) should help.
3. Internally the EntityRuler uses the Matcher and PhraseMatcher. You can think of the EntityRuler as a wrapper for the matchers that lets you use them as a pipeline component.
4. When matches overlap, [the EntityRuler uses the longer match](https://spacy.io/usage/rule-based-matching#entityruler-usage:~:text=If%20any%20matches%20were%20to%20overlap%2C,first%20in%20the%20Doc%20is%20chosen.), or the earliest one if they are equally long.

**Reply by nlovell1** [reply]
```
import spacy
from spacy.matcher import PhraseMatcher
from spacy.tokens import Doc
from spacy.pipeline import EntityRuler

terms = ["気になる", "手も足も出ない", "んだ", "手を取る"]

nlp = spacy.load("ja_core_news_lg")

doc = nlp("健太郎はアスカの手を取りながら歩いていた")

print([(token.lemma_) for token in doc])

class EntityRetokenizeComponent:
  def __init__(self, nlp):
    pass
  def __call__(self, doc):
    with doc.retokenize() as retokenizer:
        for ent in doc.ents:
            retokenizer.merge(doc[ent.start:ent.end], attrs={"LEMMA": str(doc[ent.start:ent.end])})
    return doc

ruler = EntityRuler(nlp)

ruler.add_patterns([{"label": "EXPRESSION", "pattern": term} for term in terms])
nlp.add_pipe(ruler) 

retokenizer = EntityRetokenizeComponent(nlp)
nlp.add_pipe(retokenizer, name='merge_phrases')

doc = nlp("健太郎はアスカの手を取りながら歩いていた")

print([(token.text, token.lemma_) for token in doc])

```

Thanks for the information. I have come up with a new solution. It seems that two things are happening here:
Despite the fact the lemmas for the phrase `手を取り` are `手`, `を`, and `取る`, it does not seem to reflect the term `手を取る` that I had specified.

Additionally, so far I have not been able to figure out a solution that can infer kanji/hiragana. For example, if `手を取り` was written as `手をとり`, can SpaCy accommodate for something like that?

Also, I realize this thread is a little long and has probably reached beyond the scope of this medium. Is there an alternative community that you might suggest that could be more appropriate for these inquiries?

Thank you very much for your time.

**Reply by polm** [reply]
Huh, I didn't realize you were using spaCy 2 this whole time. Is there a reason you aren't using v3?

The issue with your code is you didn't tell the PhraseMatcher in the EntityRuler to match on lemmas, you need to make this change:

    ruler = EntityRuler(nlp, phrase_matcher_attr="LEMMA")

> Additionally, so far I have not been able to figure out a solution that can infer kanji/hiragana. For example, if 手を取り was written as 手をとり, can SpaCy accommodate for something like that?

It depends on whether the tokenizer can infer the correct lemma from context. In this particular case it can. I would expect this would work most of the time in ordinary text.

> Is there an alternative community that you might suggest that could be more appropriate for these inquiries?

I'm not aware of a better place, no.

**Reply by nlovell1** [reply]
> Huh, I didn't realize you were using spaCy 2 this whole time. Is there a reason you aren't using v3?

No, it's just that I started this before the new release came out. What changes would I have to make (if any)?

So, I finally got everything to work. Here is an example of a phrase before loading the dictionary and then after (adding patterns).
```
doc = nlp("お誕生日おめでとうございます")
...
print([(token.text, token.pos_) for token in doc])
[('お', 'NOUN'), ('誕生', 'NOUN'), ('日', 'NOUN'), ('おめでとう', 'INTJ'), ('ござい', 'AUX'), ('ます', 'AUX')]
...
print([(token.text, token.pos_) for token in doc])
[('お誕生日おめでとうございます', 'AUX')]
```
What in Spacy decides the token.pos after you introduce the patterns?

In my code, I have this line:
`ruler.add_patterns([{"label": "EXPRESSION", "pattern": x.text} for x in word.kanji_forms])`

Where is the label "EXPRESSION" stored?

Thank you very much.

**Reply by polm** [reply]
Ah, I forgot this discussion started that long ago :sweat_smile: To work with spaCy 3, for this code the main thing you would have to change is how you create the pipeline component. Basically just change it to be this:

    ruler = nlp.add_pipeline("entity_ruler", ...[options go here]...)

I think that's the only change in this case.

> What in Spacy decides the token.pos after you introduce the patterns?

This is determined by the retokenizer, you can read about that [here](https://spacy.io/usage/linguistic-features#retokenization). Basically you just pass in any attributes you want to set when retokenizing.

> Where is the label "EXPRESSION" stored?

EXPRESSION is an entity label. You can check the label on individual tokens by using the `Token.entity_type_` attribute, or you can use `Doc.ents` to get all the entities in a sentence.

**Reply by nlovell1** [reply]
Recall that I wrote this code:
```
class EntityRetokenizeComponent:
  def __init__(self, nlp):
    pass
  def __call__(self, doc):
    with doc.retokenize() as retokenizer:
        for ent in doc.ents:
            retokenizer.merge(doc[ent.start:ent.end], attrs={"LEMMA": str(doc[ent.start:ent.end])})
    return doc

ruler = EntityRuler(nlp, phrase_matcher_attr="LEMMA")
```
Additionally, I added various patterns from the dictionary.
Afterwards, I ran this:
```
nlp.add_pipe(ruler) 

retokenizer = EntityRetokenizeComponent(nlp)
nlp.add_pipe(retokenizer, name='merge_phrases')
```
I understand your suggestion as far as the entity ruler goes. However, how would I accomodate for my custom retokenizer object? It seems the docs suggest to use the @Language.family decoration, but I'm a little confused how this would work.

Additionally, on an older version of my code, I used the nlp.to_disk command to serialize the model as you mentioned. I then tried to load the model in another code and it wasn't able to load all the components properly. Is there something I'm missing here?

Lastly, I'd love to help collaborate on this project to give back for the help you've given me. Are there any development opportunities that I could take part in despite being a novice?

**Reply by polm** [reply]
> I understand your suggestion as far as the entity ruler goes. However, how would I accomodate for my custom retokenizer object?

Ah, right, I forgot about that. That would be a [custom component](https://spacy.io/usage/processing-pipelines#custom-components) so you just need to register it, which would look like this:

```
from spacy.language import Language

@Language.component("entity_retokenizer")
class EntityRetokenizerComponent():
    ...

nlp.add_pipe("entity_retokenizer")
```

Also, I should have mentioned this earlier, but there's a [built-in component to merge entities](https://spacy.io/api/pipeline-functions#merge_entities) that you could also use. It wouldn't have worked when you were using the PhraseMatcher rather than the EntityRuler since it relies on the entity labels.

> Additionally, on an older version of my code, I used the nlp.to_disk command to serialize the model as you mentioned. I then tried to load the model in another code and it wasn't able to load all the components properly. Is there something I'm missing here?

What kind of error were you getting? Most components can be serialized in a straightforward way, but SudachiPy can't, so that has caused problems in the past. I think the issues with that should have been resolved but it's possible there are still some problems. 

> Lastly, I'd love to help collaborate on this project to give back for the help you've given me. Are there any development opportunities that I could take part in despite being a novice?

Thanks! One thing you could do would be to take a look at issues tagged "help wanted" or "help wanted (easy)". Another would be to add your extension to [spaCy Universe](https://spacy.io/universe) with a PR, or write a blog post about making a custom extension. Blog posts are great because they let us understand more about how people are using the library, and let other users know how easy (or hard) it is to develop extensions, and can be good at generating ideas.

**Reply by nlovell1** [reply]
```
@Language.component("entity_retokenizer")
class EntityRetokenizeComponent:
  def __init__(self, nlp):
    pass
  def __call__(self, doc):
    with doc.retokenize() as retokenizer:
        for ent in doc.ents:
            retokenizer.merge(doc[ent.start:ent.end], attrs={"LEMMA": str(doc[ent.start:ent.end])})
    return doc

ruler = nlp.add_pipeline("entity_ruler", phrase_matcher_attr="LEMMA")
...
nlp.add_pipe("entity_retokenizer", name='merge_phrases')
```
I'm still having issues here:
`AttributeError: 'Japanese' object has no attribute 'add_pipeline'`
Did I do something wrong?

**Reply by polm** [reply]
It's `add_pipe`, not `add_pipeline`.

**Reply by nlovell1** [reply]
Ah! I tried that too.
`TypeError: add_pipe() got an unexpected keyword argument 'phrase_matcher_attr'`

**Reply by polm** [reply]
Ah, that needs to go in a config object.

```
nlp.add_pipe("entity_ruler", config={"phrase_matcher_attr": "LEMMA"})
```

**Reply by nlovell1** [reply]
```
doc2 = nlp("お体大事にしてください")
print([(token.text, token.pos_) for token in doc2])
...
TypeError: 'EntityRetokenizeComponent' object is not iterable
```
What's happening here? Upgrading to 3.0 is becoming a headache! lol

**Reply by polm** [reply]
Ah, I should have been more explicit with that. There are basically two ways to register components in v3, one for simple stateless components, one for components that have state and, because of that, need a factory. Since your component has no state it can use the simple method, but that means it doesn't have to be a class anymore. 

```
@Language.component("entity_retokenizer")
def entity_retokenizer(doc):
    with doc.retokenize() as retokenizer:
        for ent in doc.ents:
            retokenizer.merge(doc[ent.start:ent.end], attrs={"LEMMA": str(doc[ent.start:ent.end])})
    return doc
```

For more details, please check [the documentation](https://spacy.io/usage/processing-pipelines#custom-components-simple).

**Reply by nlovell1** [reply]
Thanks for bearing with me on these issues here. Your fix worked. As for your earlier question, this is the error I get when I load the model.
EDIT: I did not include the custom component again in the new file, which caused the error.

Additionally, a weird behavior is occurring. When I check the output of the tokenization from before and after I add the entity ruler and entity retokenizer pipes, despite not having added any patterns(I made a new test file), the output of the phrase `一つ二つ` changes from `('一', 'NUM'), ('つ', 'NOUN'), ('二', 'NUM'), ('つ', 'NOUN')` to `('一つ', 'NUM'), ('二つ', 'NOUN')`. Is this a bug?

**Reply by polm** [reply]
> Additionally, a weird behavior is occurring.

You are merging *all* entities, not just your EXPRESSION entities. In the default Japanese models 一つ is a CARDINAL entity.

**Reply by nlovell1** [reply]
Understood. Thank you.

Something I've noticed is that the Entity Ruler is a. slow and b. gets increasingly slower the more entries are added. Is this a known effect of using the Entity Ruler to match phrases?

**Reply by polm** [reply]
The PhraseMatcher, which the EntityRuler is using internally,  should be efficient even with very large terminology lists, though there is a limit. It should not be slow in general, but in this case I guess you have maybe a few 10,000 entries, which is at the high end? You can read more about the speed of the PhraseMatcher [here](https://github.com/explosion/spaCy/issues/5999#issuecomment-685456586); that comment is for v2 but I think the internals are basically the same.

Also some things that might be issues:

1. If building the EntityRuler is slow, build it once and serialize it. 
2. Is it possible that the retokenizer is actually the slow part? If so I would recommend disabling it and just using entity labels.

**Reply by nlovell1** [reply]
That makes sense now. I was trying to index all the nouns to see what would happen, which so far as we speak has incorporated over 139 thousand entries.

In your first suggestion, do you simply mean serializing every 10k entries or so? Or something else?
Thanks again

**Reply by polm** [reply]
> 139 thousand entries.

Ah, yeah, that's too many and will be slow. The PhraseMatcher is efficient but patterns aren't free. 10k to 100k is kind of borderline but should still be usable, and I think that covers expressions.

> In your first suggestion, do you simply mean serializing every 10k entries or so? Or something else?

No, just after you've built the whole thing. The initial construction of the PhraseMatcher can be slow depending on how you build your documents, but can be skipped if you're just deserializing it. (Not sure if it's relevant to your case but we've had several issues specifically about building the PhraseMatcher being slow, so I mentioned it.)

**Reply by nlovell1** [reply]
Just to update, Spacy seems to crash at around 100k entries (and gets slower with each new entry). I think a better solution would be to just make a new tokenizer to make it work how I want, and do something you mentioned earlier with hand annotating a ton of sentences. I also want to work on sense disambiguation and I'm not sure the best way to do that, whether it be through Spacy or another tool.

**Reply by nlovell1** [reply]
Hi @polm, circling back here, and thought this might be a relevant place to ask a question I had.
To refresh, this thread was originally about adding every entry in JMDict, adding them to the PhraseMatcher. I didn't know it then, but this was sort of naive attempt to detect MWE's (which as you probably know is sort of an interesting problem to work on in Japanese).

I'm trying to do much of the same right now in a small well behaved subset of Spanish text, with reasonably good results from Wiktionary entries.

Suppose we use a basic interlingual definition that an MWE is a set of tokens that, when combined in a specific order, means something different or more specific than the basic combination of the constituent tokens.

1. If we are able to fine tune rule-based results based on grammatical phenomena and consulting a human-compiled dictionary list, how accurate can we expect to be with this naive approach? I know that this answer is dependent on available resources for dictionaries, the language in question, our definition for a MWE and whether or not we want to be able to sub-categorize different MWE types (a. la. recent literature). But I fail to understand well why the most recent literature is creating custom ML models and fine tuning them to get quite low accuracy in MWE detection.

2. In somewhat of a response to (1.), I realize that new MWEs appear in language all the time- perhaps faster than can be recorded by voluntary edits to Wiktionary, or perhaps it's a language whose lexicon isn't extensively documented. How can we use SpaCy to fill this gap, and, is this a problem that can be solved with enough training data? What bottlenecks are we working with given the state of the art computation available to a normal person like me?

3. Furthermore, what is the most comprehensive way for someone like myself to be familiar with the state of the art for MWE and other NLP tasks alike, and understand the methods used?

Thank you very much for your time.

**Reply by polm** [reply]
I'm afraid I'm not really familiar enough with MWE research to answer your first few questions. My impression is that finding a useful definition of MWEs that covers many use cases is very hard, and that finding emerging phrases is a small but distinct field of research.

 For understanding research better and learning more about it, I'd recommend:

1. Check [Jurafsky & Martin](https://web.stanford.edu/~jurafsky/slp3/) for an overview
2. Check [NLP Progress](http://nlpprogress.com/) or [Papers with Code](https://paperswithcode.com/) for benchmarks and to see trends
3. Check [Semantic Scholar](https://www.semanticscholar.org/) for specific papers, and check for citations of interest

**Reply by nlovell1** [reply]
Hi, and thanks again. I appreciate your help on this thread, it's been very informative and helpful for my development. Have a good day.

**Reply by lawctan** [reply]
Hi @nlovell1, I'm curious if you have you been able to solve this issue for japanese parsing quality? Did you end up just doing the retokinization approach? Thanks!

**Reply by nlovell1** [reply]
Hi @lawctan , it's my impression that detecting multi word expressions in general is not something that's of great interest in the field. Some attempts have limited success (Babieno 2022, for metaphors for example) but accuracy never breaks 70 percent. I would suggest rethinking your use case (why do you need to detect MWEs)?

This problem is likely outside the scope of Spacy.

Let me know if you have any questions.

**Reply by lawctan** [reply]
it's basically to achieve something similar to what jisho.org can do. For example, in https://jisho.org/search/%E6%98%A8%E6%97%A5%E3%81%99%E3%81%8D%E7%84%BC%E3%81%8D%E3%82%92%E9%A3%9F%E3%81%B9%E3%81%BE%E3%81%97%E3%81%9F, it's able to know that 食べました belongs together.

**Reply by lawctan** [reply]
Eventually, i went with an approach where I looked at the dep and pos fields of the neighboring tokens of a verb to retokenize them

**Reply by lawctan** [reply]
Actually, just realized that the original post is about a word that was treated as a noun even though it should have been treated as part of a verb
